rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,393 @@
+
+## API
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  These requests are used
+to transmit metadata of an HTTP request which the service requires the client
+to perform.
+
+An example usage of this format is uploading a file.  The creation endpoint of
+this service will generate all of the HTTP request metadata that the uploading
+process will use to upload file.  Then the uploader will match these HTTP
+requests with payload bodies and run each of these requests.
+
+The motivation for this type of flow is to support the case where the uploader
+is running in a different region than this service is.  If we required all
+uploads to gate through this service, we'd need to duplicate uploading effort
+and perform unneeded inter-region transfers.  By transmitting a list of
+requests which allow the uploading to happen, we can instruct the uploader to
+upload directly to the existing S3 infrastructure, but perform all signing
+within this service.
+
+A generalised request will look like:
+
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in
+  the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key ->
+  list of strings)
+* All query string, path, fragments are contained in a fully formed `url`
+  property
+
+### Origins
+Where present, the `origin` rest parameter in this service will be either an
+IPv4 address, IPv6 address, or an identifier.  This parameter will specify the
+source location of the ultimately originating request to upload.  If the origin
+is an IPv4 or IPv6, a mapping of the address and IP will occur to find the
+backing storage.  If the origin is not an IPv4 or IPv6 address and is identical
+to a set of known and mappable identifiers (e.g.  `s3_us-west-2`), then that
+will be used to map the token to find the backing storage provider.  If the
+origin is not an IP address and not a configured identifier, an error will be
+sent back to the client.
+
+The mapping of origins to storage providers should be statically configured.
+
+The origin `all` is reserved and cannot be used.  This is to enable management
+endpoints to operate on all origins in the future.
+
+### Creation
+
+```
+PUT /objects/:name
+PATCH /objects/:name
+```
+
+The `PUT` endpoint will take an origin and name.  It will return a `200` with a
+document that contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the object in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
```

Comment:
I'm assuming `content` and `transfer` here map to raw content and encoded content.

First, the differences need to be called out explicitly. In HTTP spec terms, this is the *representation* of content and language needs to clarify how this maps to `Content-Type`, `Content-Encoding`, and `Transfer-Encoding`. I'm guessing/hoping that `Transfer-Encoding` doesn't come into play and that `content*` and `transfer*` as they are defined here map to `Content-Type` and `Content-Encoding` representations, respectively. I say I'm hoping `Transfer-Encoding` doesn't come into play because it is a hop-by-hop header and there's no guarantee it will be preserved across proxies, etc. It is also the lowest level encoding and HTTP libraries typically transparently filter HTTP message bodies through `Transfer-Encoding` and there's often not a way to access the raw bytes going over the unencrypted wire.

The use of *transfer* when `Transfer-Encoding` is used in the context of HTTP encoding can be a bit confusing. So it might be worth finding a better name to avoid the ambiguity.

Reply:
Yes, content and transfer in this context refer to the actual thing you asked our library to upload (i.e. before content-encoding was applied) and the transfer refers to the thing which was transferred (i.e. the content-encoding-encoded version).

the distinction here is between the content of the document and the bytes which were actually transferred.

There's no intention to use `Transfer-Encoding`.

```
    res.set('x-taskcluster-location-content-sha256', artifact.details.contentSha256);
    res.set('x-taskcluster-location-content-length', artifact.details.contentLength);
    res.set('x-taskcluster-location-transfer-sha256', artifact.details.transferSha256);
    res.set('x-taskcluster-location-transfer-length', artifact.details.transferLength);
    res.set('x-taskcluster-location-content-encoding', artifact.details.contentEncoding || 'identity');
    res.set('x-taskcluster-location-content-type', artifact.details.contentType);
```
This is how the headers map to these values.

Yes, `transfer` might be confusing, but I can't think of a better name to describe the sha256 of the bytes which were actually transfered... Do you have any ideas?  We also already have that name in the blob api, so it would be a decent amount of work to change.

Reply:
How about objectLength/Sha256 and contentLength/Sha256?  If I understand the above correctly, then contentLength would correspond exactly to the content-length header, which seems to make sense.  I'm not confident I understand the above :)